src/ClassDiscovery.php

```diff
@@ -2,126 +2,117 @@
 
 namespace Http\Discovery;
 
-use Puli\Discovery\Api\Discovery;
+use Http\Discovery\Exception\DiscoveryFailedException;
+use Http\Discovery\Exception\StrategyUnavailableException;
 
 /**
  * Registry that based find results on class existence.
  *
  * @author David de Boer <david@ddeboer.nl>
  * @author Márk Sági-Kazár <mark.sagikazar@gmail.com>
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
  */
 abstract class ClassDiscovery
 {
     /**
-     * @var GeneratedPuliFactory
+     * A list of strategies to find classes.
+     *
+     * @var array
      */
-    private static $puliFactory;
+    public static $strategies = [
+        Strategy\Puli::class,
+        Strategy\HttpClients::class,
+        Strategy\GuzzleFactory::class,
+        Strategy\DiactorosFactory::class,
+    ];
 
     /**
-     * @var Discovery
+     * Discovery cache to make the second time we use discovery faster.
+     *
+     * @var array
      */
-    private static $puliDiscovery;
+    private static $cache = [];
 
     /**
-     * @return GeneratedPuliFactory
+     * Finds a class.
+     *
+     * @param string $type
+     *
+     * @return string
+     *
+     * @throws DiscoveryFailedException
      */
-    public static function getPuliFactory()
+    protected static function findOneByType($type)
     {
-        if (null === self::$puliFactory) {
-            if (!defined('PULI_FACTORY_CLASS')) {
-                throw new \RuntimeException('Puli Factory is not available');
-            }
-
-            $puliFactoryClass = PULI_FACTORY_CLASS;
+        // Look in the cache
+        if (null !== $class = self::getFromCache($type)) {
```

Comment:
Although not necessary, I would use this: `null !== ($class = self::getFromCache($type))` More readable I think